src/librustc_mir/const_eval/mod.rs

```diff
@@ -67,3 +69,32 @@ pub(crate) fn destructure_const<'tcx>(
 
     mir::DestructuredConst { variant, fields }
 }
+
+pub(crate) fn deref_const<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    val: &'tcx ty::Const<'tcx>,
+) -> &'tcx ty::Const<'tcx> {
+    trace!("deref_const: {:?}", val);
+    let ecx = mk_eval_cx(tcx, DUMMY_SP, param_env, false);
+    let op = ecx.eval_const_to_op(val, None).unwrap();
+    let mplace = ecx.deref_operand(op).unwrap();
+
+    let ty = match mplace.meta {
+        MemPlaceMeta::None => mplace.layout.ty,
+        MemPlaceMeta::Poison => bug!("poison metadata in `deref_const`: {:#?}", mplace),
+        // In case of unsized types, figure out the real type behind.
+        MemPlaceMeta::Meta(scalar) => match mplace.layout.ty.kind {
+            ty::Dynamic(..) => ecx.read_drop_type_from_vtable(scalar).unwrap().1,
+            ty::Str => bug!("there's no sized equivalent of a `str`"),
+            ty::Slice(elem_ty) => tcx.mk_array(elem_ty, scalar.to_machine_usize(&tcx).unwrap()),
+            _ => bug!(
+                "type {} should not have metadata, but had {:?}",
+                mplace.layout.ty,
+                mplace.meta
+            ),
```

Comment:
I don't see this using `struct_tail`, so it presumably will fail on `Newtype<[T]>`?

Reply:
jup, that's impossible to do in the general case, as we may have just nongeneric `struct NewType(i32, [u8]);` and there's no sized equivalent.

Reply:
"should not have metadata" isn't true though for `NewType<[T]>` or `dyn Trait`. So maybe adjust the `bug!` message (and then it can be merged with `Str`).

Also the query docs and the doc comment on this function should clarify that the only supported unsized type is arrays.